sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,411 @@
+                return ret
+
+        else:
+            # Here we may safely assume HalfMultinomialLoss aka categorical
+            # cross-entropy.
+            # HalfMultinomialLoss computes only the diagonal part of the hessian, i.e.
+            # diagonal in the classes. Here, we want the matrix-vector product of the
+            # full hessian. Therefore, we call gradient_proba.
+            gradient, proba = self._loss.gradient_proba(
+                y_true=y,
+                raw_prediction=raw_prediction,
+                sample_weight=sample_weight,
+                n_threads=n_threads,
+            )
+            grad = np.empty((n_classes, n_dof), dtype=X.dtype, order="F")
+            grad[:, :n_features] = gradient.T @ X + l2_reg_strength * w
+            if self.fit_intercept:
+                grad[:, -1] = gradient.sum(axis=0)
+
+            # Full hessian-vector product, i.e. not only the diagonal part of the
+            # hessian. Derivation with some index battle for input vector s:
+            #   - sample index i
+            #   - feature indices j, m
+            #   - class indices k, l
+            #   - 1_{k=l} is one if k=l else 0
+            #   - p_i_k is the (predicted) probability that sample i belongs to class k
+            #     for all i: sum_k p_i_k = 1
+            #   - s_l_m is input vector for class l and feature m
+            #   - X' = X transposed
+            #
+            # Note: Hessian with dropping most indices is just:
+            #       X' @ p_k (1(k=l) - p_l) @ X
+            #
+            # result_{k j} = sum_{i, l, m} Hessian_{i, k j, m l} * s_l_m
+            #   = sum_{i, l, m} (X')_{ji} * p_i_k * (1_{k=l} - p_i_l)
+            #                   * X_{im} s_l_m
+            #   = sum_{i, m} (X')_{ji} * p_i_k
+            #                * (X_{im} * s_k_m - sum_l p_i_l * X_{im} * s_l_m)
+            #
+            # See also https://github.com/scikit-learn/scikit-learn/pull/3646#discussion_r17461411  # noqa
+            def hessp(s):
+                s = s.reshape((n_classes, -1), order="F")  # shape = (n_classes, n_dof)
+                if self.fit_intercept:
+                    s_intercept = s[:, -1]
+                    s = s[:, :-1]
+                else:
+                    s_intercept = 0
+                tmp = X @ s.T + s_intercept  # X_{im} * s_k_m
+                tmp += (-proba * tmp).sum(axis=1)[:, np.newaxis]  # - sum_l ..
+                tmp *= proba  # * p_i_k
+                if sample_weight is not None:
+                    tmp *= sample_weight[:, np.newaxis]
+                hess_prod = np.empty_like(grad, order="F")
+                hess_prod[:, :n_features] = tmp.T @ X + l2_reg_strength * s
+                if self.fit_intercept:
+                    hess_prod[:, -1] = tmp.sum(axis=0)
+                if coef.ndim == 1:
+                    return hess_prod.ravel(order="F")
+                else:
+                    return hess_prod
```

Comment:
codecov seems to imply that there are not test for the newton-cg + multinomial multiclass combo.

Reply:
I don't understand, this line should be covered by `test_multinomial_grad_hess`, no?

Reply:
Not yet, I'll extend the test to do so.

Reply:
Ok even for `newton-cg` we do ravel `w0`. So this code has no chance to ever be called this way. Should't we just remove this branch and update the docstring?

Reply:
If we go for a IRLS solver, which is a good choice for n_samples >> n_features, this branch can be used. I like the idea of keeping the shape of the passed `coef`. But I can remove this, if you prefer.

Reply:
No this is fine, the code is not too complex. Thanks for adding the missing unit-tests to cover this branch.